woocommerce/woocommerce
# 25021 add pay order product has enough stock filter

### All Submissions:


### Changes proposed in this Pull Request:

Adds a filter within the `WC_Shortcode_Checkout::order_pay()` method after the `woocommerce_pay_order_product_in_stock` filterable check, so that customer order statuses awaiting payment can still be paid as some products may have had already stock lowered in some circumstances and they don't need these checks anymore. (Applies to Order Status Manager).

Closes #25021 (see issue for more details)

### How to test the changes in this Pull Request:

1. Have a customer process a complete checkout with an item in stock
2. Customer should be able to process the checkout without errors

#### Scenario 2

1. Have an order awaiting payment: the order should contain an item with plenty of stock
2. Have customer pay for it from My Account > Orders > Pay
3. Result: order should be paid and no error should occur

#### Scenario 3

This scenario is the actual scenario the PR is going to address.

1. Install & Activate WooCommerce Order Status Manager
2. Create a custom order status with a "Payment required" flag
3. Create an inventory-managed product with stock of 1
4. Create an order using any payment method
5. Change the status of that order to any custom order status with "payment required"
6. With the customer user of that order, go to "My Account > Orders" and attempt to "Pay" that order

Currently, this would fail, because the stock has already dropped by 1 unit, and WooCommerce is attempting to drop it again; since it is at 1, it will fail and raise an user notice, preventing to complete payment. However, with this filter in place, one should be able to skip the checks in `WC_Shortcode_Checkout::order_pay()` and have the customer pay for the order awaiting payment.

### Other information:


### Changelog entry

I did not add the following in changelog.txt, just a suggestion:

> * Dev - Add filter to tweak whether a product has enough stock while attempting to pay for an order
